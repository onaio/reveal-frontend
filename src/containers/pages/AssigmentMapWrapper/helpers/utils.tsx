import { Dictionary } from '@onaio/utils';
import { Feature } from '@turf/turf';
import { History } from 'history';
import { EventData, Map } from 'mapbox-gl';
import { AssignmentMapWrapperProps } from '..';
import { PLAN_TYPES_WITH_MULTI_JURISDICTIONS } from '../../../../configs/env';
import { getPlanType } from '../../../../helpers/utils';
import { SELECTION_REASON } from '../../../../store/ducks/opensrp/hierarchies/constants';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/**
 * Handler to get current parent id from clicked jurisdiction on map
 * @param props - AssignmentMapWrapper component props
 * @param setMapParent - function to set map parent id to state
 * @param history - history session object
 */

export const onJurisdictionClick = (
  props: AssignmentMapWrapperProps,
  setMapParent: (currentId: string) => void,
  history: History
) => {
  function handleJurisdictionClick(_: Map, e: EventData) {
    // Destructure event data
    const { point, target, originalEvent } = e;
    const {
      currentParentNode,
      rootJurisdictionId,
      selectNodeCreator,
      deselectNodeCreator,
      currentChildren,
      plan,
      baseAssignmentURL,
    } = props;
    // grab underlying features from map
    const features: Feature[] = target.queryRenderedFeatures(point);
    if (!features.length) {
      return false;
    }

    for (const feature of features) {
      const activeJurisdictionId =
        (feature && feature.id) ||
        (feature && feature.properties && feature.properties.jurisdiction_id);
      if (activeJurisdictionId) {
        const currentId = activeJurisdictionId;
        let activeCurrentNode: TreeNode | undefined;

        if (currentParentNode) {
          // ensure currentparentnode is not undefined first
          // Handle selection for admin level 0
          if (
            currentParentNode.model.id === rootJurisdictionId &&
            currentId === currentParentNode.model.id
          ) {
            activeCurrentNode = currentParentNode;
          } else {
            // Handle selection for admin level > 0
            activeCurrentNode = currentChildren.find(
              (node: TreeNode) => node.model.id === currentId
            );
          }
        }
        const thePlanType = getPlanType(plan);
        if (
          activeCurrentNode &&
          originalEvent.altKey &&
          PLAN_TYPES_WITH_MULTI_JURISDICTIONS.includes(thePlanType)
        ) {
          if (!nodeIsSelected(activeCurrentNode)) {
            selectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          } else {
            deselectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          }
          setMapParent(currentId);
        } else {
          if (activeCurrentNode && activeCurrentNode.hasChildren()) {
            history.push(`${baseAssignmentURL}/${currentId}`);
          }
        }
        break;
      }
    }
  }
  return handleJurisdictionClick;
};

/**
 * Map component popup handler
 * @param map - mapbox map instance
 * @param event - map event object
 */

export const buildMouseMoveHandler = (map: Map, event: EventData) => {
  const popup = new mapboxgl.Popup({
    closeButton: false,
  });
  // Added this part to remove duplicate popup containers showing on map when hovering
  const nodes: any = document.getElementsByClassName('mapboxgl-popup');
  let node: HTMLDivElement;
  if (nodes && nodes.length) {
    for (node of nodes) {
      if (node.style) {
        node.style.display = 'none';
      }
    }
  }
  let content: string = '';
  let feature: Feature | any = {};
  popup.remove();
  // grab underlying features from map
  const features: any = event.target.queryRenderedFeatures(event.point) as Dictionary[];
  //
  if (!features.length) {
    return;
  }

  // Change the cursor style to pointer if features exist.
  map.getCanvas().style.cursor = features.length ? 'pointer' : 'grab';

  // loop through features, setting the name property to content var
  for (feature of features) {
    if (feature && feature.properties && feature.properties.name) {
      content += `<div class="jurisdiction-name"><center>${feature.properties.name}</center></div>`;
    }
    /**
     * Break out of loop once content is set
     * This helps in a case where we could have 2 or more identical features
     */
    if (content) {
      break;
    }
  }
  /**
   * Add popup to map after content is set
   */
  if (content.length) {
    popup
      .setLngLat(map.unproject(event.point))
      .setHTML(content)
      .addTo(map);
  } else {
    // clear existing popups if content is empty
    popup.remove();
    return;
  }
  return true;
};
