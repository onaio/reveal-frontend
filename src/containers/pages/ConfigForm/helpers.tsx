import React from 'react';
import Loading from '../../../components/page/Loading';
import { NoDataComponent } from '../../../components/Table/NoDataComponent';
import { OPENSRP_API_BASE_URL } from '../../../configs/env';
import { getAcessTokenOrRedirect, growl } from '../../../helpers/utils';
import { newGetPayloadOptions } from '../../../services/opensrp';

/** drill down props */
export const drillDownProps = {
  paginate: false,
  renderNullDataComponent: () => <NoDataComponent />,
};

export const defaultConfigProps = {
  LoadingComponent: <Loading />,
  accessToken: getAcessTokenOrRedirect,
  baseURL: OPENSRP_API_BASE_URL,
  customAlert: growl,
  getPayload: newGetPayloadOptions,
};
