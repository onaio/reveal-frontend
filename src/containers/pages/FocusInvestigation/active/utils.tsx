import React from 'react';
import { RouteComponentProps } from 'react-router';
import { Column } from 'react-table';
import DrillDownTableLinkedCell from '../../../../components/DrillDownTableLinkedCell';
import {
  defaultOptions,
  renderInFilterFactory,
} from '../../../../components/Table/DrillDownFilters/utils';
import { NoDataComponent } from '../../../../components/Table/NoDataComponent';
import { NO_INVESTIGATIONS_FOUND } from '../../../../configs/lang';
import { extractPlan, removeNullJurisdictionPlans } from '../../../../helpers/utils';
import { OpenSRPService } from '../../../../services/opensrp';
import { Plan } from '../../../../store/ducks/plans';

/** construct props for DrillDown for activeFI page
 * Generic type D is the type of data that will be passed to the Table
 * its not `Plan` because the plan goes through a few transformations before being rendered,
 * at render its structure will have mutated to a Dictionary courtesy of 'extractPlan'
 * @param columns - react-table compatible columns
 * @param data - array of data to be rendered in table
 * @param queryParam - the key to be used by searchForm to use as key in searchParam
 */
export const createTableProps = <D extends object, TProps extends RouteComponentProps>(
  columns: Array<Column<D>>,
  data: Plan[] | null,
  props: TProps,
  paramString: string,
  service: typeof OpenSRPService,
  noDataMessage: string = NO_INVESTIGATIONS_FOUND
) => {
  const cleanedData = data !== null ? data : [];
  const jurisdictionValidPlans = removeNullJurisdictionPlans(cleanedData);
  const thePlans = jurisdictionValidPlans.map(extractPlan);

  return {
    CellComponent: DrillDownTableLinkedCell,
    columns,
    data: thePlans,
    renderInBottomFilterBar: renderInFilterFactory({
      showColumnHider: false,
      showFilters: false,
      showPagination: true,
      showRowHeightPicker: false,
      showSearch: false,
    }),
    renderInTopFilterBar: renderInFilterFactory({
      ...defaultOptions,
      componentProps: props,
      queryParam: paramString,
      serviceClass: service,
    }),
    renderNullDataComponent: () => <NoDataComponent message={noDataMessage} />,
    useDrillDown: false,
  };
};
