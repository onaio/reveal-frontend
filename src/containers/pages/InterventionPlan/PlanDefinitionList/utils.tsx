import { DrillDownColumn } from '@onaio/drill-down-table';
import { Dictionary } from '@onaio/utils/dist/types/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell, Row } from 'react-table';
import { isFIOrDynamicFI } from '../../../../components/forms/PlanForm/helpers';
import { DISPLAYED_PLAN_TYPES } from '../../../../configs/env';
import {
  CASE_TRIGGERED_TITLE,
  FOCUS_INVESTIGATION_STATUS_REASON,
  INTERVENTION_TYPE_LABEL,
  LAST_MODIFIED,
  ROUTINE_TITLE,
  STATUS_HEADER,
  TITLE,
} from '../../../../configs/lang';
import { planStatusDisplay, UseContext } from '../../../../configs/settings';
import {
  CASE_TRIGGERED,
  FI_REASON_CODE,
  INTERVENTION_TYPE_CODE,
  PLAN_UPDATE_URL,
  ROUTINE,
} from '../../../../constants';
import { InterventionType, PlanRecord } from '../../../../store/ducks/plans';

/**
 * sort plan UseContext values
 * @param {Row<PlanRecord>} rowA - table props
 * @param {Row<PlanRecord>} rowB - table props
 * @param {string} code - Plan use context code
 */
const sortUseContext = (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>, code: string) => {
  /** custom sort function based useContext */
  const getIntervention = (row: Row<PlanRecord>) => {
    const interventionPlanContext = (row.original as Dictionary).plan_useContext.filter(
      (context: UseContext) => context.code === code
    );
    return interventionPlanContext.length > 0 ? interventionPlanContext[0].valueCodableConcept : '';
  };
  return getIntervention(rowA) > getIntervention(rowB) ? 1 : -1;
};

/**
 *
 * @param {UseContext[]} value - List of plan use contexts
 * @param {string} code - Plan use context code
 */
const getUseContext = (value: UseContext[], code: string) => {
  const typeUseContext = value.filter((e: any) => e.code === code);
  if (typeUseContext[0]?.valueCodableConcept === CASE_TRIGGERED) {
    return CASE_TRIGGERED_TITLE;
  }
  if (typeUseContext[0]?.valueCodableConcept === ROUTINE) {
    return ROUTINE_TITLE;
  }
  return typeUseContext.length > 0 ? typeUseContext[0].valueCodableConcept : null;
};

// plan intervention type column
const interventionTypeColumn = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, INTERVENTION_TYPE_CODE),
  Header: INTERVENTION_TYPE_LABEL,
  accessor: 'plan_useContext',
  id: 'plan_use_context_intervention_type',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, INTERVENTION_TYPE_CODE),
} as DrillDownColumn<PlanRecord>;

// plan FI reason column
const fiReasonColumns = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, FI_REASON_CODE),
  Header: FOCUS_INVESTIGATION_STATUS_REASON,
  accessor: 'plan_useContext',
  id: 'plan_use_context_fi_reason',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, FI_REASON_CODE),
} as DrillDownColumn<PlanRecord>;

const displayedPlans = [...DISPLAYED_PLAN_TYPES] as InterventionType[];
const showFiReasonColumns =
  displayedPlans.length === 2 &&
  isFIOrDynamicFI(displayedPlans[0]) &&
  isFIOrDynamicFI(displayedPlans[1]);

const interventionOrFiReasonColumn = showFiReasonColumns ? fiReasonColumns : interventionTypeColumn;

export const TableColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <Link to={`${PLAN_UPDATE_URL}/${original.id}`} key={original.id}>
          {cell.value}
        </Link>
      );
    },
    Header: TITLE,
    accessor: 'plan_title',
    minWidth: 200,
  },
  interventionOrFiReasonColumn,
  {
    Cell: (cell: Cell<PlanRecord>) => {
      return planStatusDisplay[cell.value] || null;
    },
    Header: STATUS_HEADER,
    accessor: 'plan_status',
    maxWidth: 30,
    minWidth: 10,
  },
  {
    Header: LAST_MODIFIED,
    accessor: 'plan_date',
    maxWidth: 30,
    minWidth: 20,
  },
];
