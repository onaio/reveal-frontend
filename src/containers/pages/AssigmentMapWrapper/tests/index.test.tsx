import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import { EventData, Popup } from 'mapbox-gl';
import * as React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { AssignmentMapWrapper, ConnectedAssignmentMapWrapper } from '..';
import { getJurisdictions } from '../../../../components/TreeWalker/helpers';
import { MAP_LOAD_ERROR } from '../../../../configs/lang';
import { PlanDefinition } from '../../../../configs/settings';
import { ASSIGN_PLAN_URL, MANUAL_ASSIGN_JURISDICTIONS_URL } from '../../../../constants';
import { OpenSRPService } from '../../../../services/opensrp';
import store from '../../../../store';
import hierachyReducer, {
  deselectNode,
  fetchTree,
  getCurrentChildren,
  getMetaData,
  reducerName as hierachyReducerName,
  selectNode,
} from '../../../../store/ducks/opensrp/hierarchies';
import { sampleHierarchy } from '../../../../store/ducks/opensrp/hierarchies/tests/fixtures';
import jurisdictionsReducer, {
  fetchJurisdictions,
  reducerName as jurisdictionsReducerName,
  removeJurisdictions,
} from '../../../../store/ducks/opensrp/jurisdictions';
import jurisdictionMetadataReducer, {
  reducerName as jurisdictionMetadataReducerName,
} from '../../../../store/ducks/opensrp/jurisdictionsMetadata';
import plansReducer, {
  reducerName as plansReducerName,
} from '../../../../store/ducks/opensrp/PlanDefinition';
import { buildMouseMoveHandler, onJurisdictionClick } from '../helpers/utils';
import * as fixtures from './fixtures';

// tslint:disable-next-line: no-var-requires
const fetch = require('jest-fetch-mock');

reducerRegistry.register(jurisdictionsReducerName, jurisdictionsReducer);
reducerRegistry.register(hierachyReducerName, hierachyReducer);
reducerRegistry.register(jurisdictionMetadataReducerName, jurisdictionMetadataReducer);
reducerRegistry.register(plansReducerName, plansReducer);

const history = createBrowserHistory();

jest.mock('../../../../components/GisidaLite', () => {
  const MemoizedGisidaLiteMock = () => <div>I love oov</div>;
  return {
    MemoizedGisidaLite: MemoizedGisidaLiteMock,
  };
});
jest.mock('../../../../configs/env', () => ({
  PLAN_TYPES_WITH_MULTI_JURISDICTIONS: ['IRS'],
}));

describe('containers/pages/AssigmentMapWrapper', () => {
  /** renders correctly, a full render cycle
   *  check that errors are raised.
   * works correctly wth store
   */
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    fetchMock.mockClear();
    fetchMock.resetMocks();
  });

  afterEach(() => store.dispatch(removeJurisdictions()));

  it('renders successfully', () => {
    shallow(
      <Router history={history}>
        <AssignmentMapWrapper />
      </Router>
    );
  });
  it('works correctly with store', async () => {
    fetch.mockResponseOnce(JSON.stringify([fixtures.jurisdiction1]), { status: 200 });
    store.dispatch(fetchJurisdictions(fixtures.payload as any));
    const props = {
      currentParentId: '2492',
      plan: {
        identifier: '2493',
      } as PlanDefinition,
      rootJurisdictionId: '2492',
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedAssignmentMapWrapper {...props} />
        </Router>
      </Provider>
    );
    const params = {
      is_jurisdiction: true,
      return_geometry: true,
    };
    const result = await getJurisdictions(['2492'], params, 1);
    await flushPromises();
    expect(fetch.mock.calls).toEqual([
      [
        'https://test.smartregister.org/opensrp/rest/location/findByJurisdictionIds?is_jurisdiction=true&return_geometry=true&jurisdiction_ids=2492',
        {
          headers: {
            accept: 'application/json',
            authorization: 'Bearer null',
            'content-type': 'application/json;charset=UTF-8',
          },
          method: 'GET',
        },
      ],
    ]);
    expect(fetch.mock.calls).toHaveLength(1);
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(wrapper.find('AssignmentMapWrapper').props()).toEqual({
      autoSelectNodesActionCreator: expect.any(Function),
      baseAssignmentURL: '/',
      currentChildren: [],
      currentParentId: '2492',
      currentParentNode: undefined,
      deselectNodeCreator: expect.any(Function),
      fetchJurisdictionsActionCreator: expect.any(Function),
      getJurisdictionsFeatures: {
        features: [],
        type: 'FeatureCollection',
      },
      getJurisdictionsMetadata: {},
      hideBottomBreadCrumbCallback: expect.any(Function),
      jurisdictionsChunkSize: 30,
      plan: {
        identifier: '2493',
      },
      rootJurisdictionId: '2492',
      selectNodeCreator: expect.any(Function),
      serviceClass: OpenSRPService,
    });
    expect(result).toEqual({ error: null, value: [fixtures.jurisdiction1] });
    wrapper.unmount();
  });
  it('test that jurisdictions are fetched correctly', async () => {
    fetch.mockResponseOnce(JSON.stringify([fixtures.jurisdiction1]), { status: 200 });
    store.dispatch(fetchTree(sampleHierarchy, '2942'));
    const children = getCurrentChildren()(store.getState(), {
      currentParentId: '2942',
      leafNodesOnly: true,
      planId: '2942',
      rootJurisdictionId: '2942',
    });
    const props = {
      currentChildren: [children] as any,
      currentParentId: '3019',
      fetchJurisdictionsActionCreator: fetchJurisdictions,
      getJurisdictionsFeatures: fixtures.geoCollection as any,
      plan: {
        identifier: '2493',
      } as PlanDefinition,
      rootJurisdictionId: '2942',
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedAssignmentMapWrapper {...props} />
        </Router>
      </Provider>
    );
    const params = {
      is_jurisdiction: true,
      return_geometry: true,
    };
    const result = await getJurisdictions(['2942'], params, 20);
    await flushPromises();
    expect(fetch).toHaveBeenCalledTimes(2);
    expect(wrapper.find('AssignmentMapWrapper').props()).toEqual({
      autoSelectNodesActionCreator: expect.any(Function),
      baseAssignmentURL: '/',
      currentChildren: expect.any(Array),
      currentParentId: '3019',
      currentParentNode: expect.any(Object),
      deselectNodeCreator: expect.any(Function),
      fetchJurisdictionsActionCreator: expect.any(Function),
      getJurisdictionsFeatures: {
        features: [],
        type: 'FeatureCollection',
      },
      getJurisdictionsMetadata: {},
      hideBottomBreadCrumbCallback: expect.any(Function),
      jurisdictionsChunkSize: 30,
      plan: {
        identifier: '2493',
      },
      rootJurisdictionId: '2942',
      selectNodeCreator: expect.any(Function),
      serviceClass: OpenSRPService,
    });
    expect(result).toEqual({ error: null, value: expect.any(Array) });
    wrapper.unmount();
  });
  it('tests that geojson is loaded correctly in store', async () => {
    fetch.mockResponseOnce(JSON.stringify([fixtures.jurisdiction1]), { status: 200 });
    store.dispatch(fetchTree(sampleHierarchy, '2942'));
    store.dispatch(fetchJurisdictions(fixtures.payload as any));
    store.dispatch(selectNode('2942', '2942', '2943'));
    const children = getCurrentChildren()(store.getState(), {
      currentParentId: '2942',
      leafNodesOnly: true,
      planId: '2953',
      rootJurisdictionId: '2942',
    });
    const props = {
      currentChildren: [children] as any,
      currentParentId: '3019',
      fetchJurisdictionsActionCreator: fetchJurisdictions,
      getJurisdictionsFeatures: fixtures.geoCollection as any,
      plan: {
        identifier: '2493',
      } as PlanDefinition,
      rootJurisdictionId: '2942',
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedAssignmentMapWrapper {...props} />
        </Router>
      </Provider>
    );
    const params = {
      is_jurisdiction: true,
      return_geometry: true,
    };
    const result = await getJurisdictions(['2942'], params, 20);
    await flushPromises();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(wrapper.find('AssignmentMapWrapper').props()).toEqual({
      autoSelectNodesActionCreator: expect.any(Function),
      baseAssignmentURL: '/',
      currentChildren: expect.any(Array),
      currentParentId: '3019',
      currentParentNode: expect.any(Object),
      deselectNodeCreator: expect.any(Function),
      fetchJurisdictionsActionCreator: expect.any(Function),
      getJurisdictionsFeatures: {
        features: [fixtures.geoCollection.features[1]],
        type: 'FeatureCollection',
      },
      getJurisdictionsMetadata: {
        '2942': {
          '2943': {
            '2942': {
              actionBy: 'User Change',
              selected: true,
            },
            '3019': {
              actionBy: 'User Change',
              selected: true,
            },
            '3951': {
              actionBy: 'User Change',
              selected: true,
            },
          },
        },
      },
      hideBottomBreadCrumbCallback: expect.any(Function),
      jurisdictionsChunkSize: 30,
      plan: {
        identifier: '2493',
      },
      rootJurisdictionId: '2942',
      selectNodeCreator: expect.any(Function),
      serviceClass: OpenSRPService,
    });
    expect(toJson(wrapper.find('MemoizedGisidaLiteMock div'))).toMatchSnapshot(
      'map renders correctly'
    );
    expect(result).toEqual({ error: null, value: expect.any(Array) });
    wrapper.unmount();
  });
  it('handles jurisdiction click on map', async () => {
    store.dispatch(selectNode('8fb28715-6c80-4e2c-980f-422798fe9f41', '2942', '3019'));
    await flushPromises();
    const event = {
      originalEvent: {
        altKey: true,
      },
      point: {
        x: 463.5,
        y: 477.1875,
      },
      target: {
        queryRenderedFeatures: () => {
          return [
            {
              geometry: {
                coordinates: [
                  [
                    [101.16072535514832, 15.119824869285075],
                    [101.15796539300004, 15.052626968000027],
                    [101.16026588800008, 15.052683043000059],
                    [101.16249336800007, 15.05279176700003],
                    [101.16457351200006, 15.05287588400006],
                    [101.16817184600006, 15.053010455000049],
                    [101.16992435000009, 15.05309115600005],
                    [101.16997576800009, 15.052226568000037],
                    [101.17015797000005, 15.050841579000064],
                    [101.17258921900009, 15.050799469000026],
                    [101.17424499900005, 15.050698872000055],
                    [101.17613469500009, 15.050485078000065],
                    [101.17699562400009, 15.050437232000036],
                  ],
                ],
                type: 'Polygon',
              },
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              properties: {
                geographicLevel: 3,
                name: 'Two Three Two Release Village',
                parentId: '872cc59e-0bce-427a-bd1f-6ef674dba8e2',
                status: 'Active',
                version: 0,
              },
              type: 'Feature',
            },
          ];
        },
      },
    };
    const mockMapObj: any = {};
    const selectNodeCreatorMock = jest.fn(selectNode);
    const deselectNodeCreatorMock = jest.fn(deselectNode);
    const mockProps = {
      currentChildren: [
        {
          model: {
            id: '2942',
            meta: { selected: false },
          },
        },
        {
          model: {
            id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
            meta: { selected: false },
          },
        },
      ],
      currentParentNode: {
        children: [
          {
            model: {
              id: '2942',
              meta: { selected: false },
            },
          },
          {
            model: {
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              meta: { selected: false },
            },
          },
        ],
        model: {
          id: '2943',
          meta: { selected: false },
        },
      },
      deselectNodeCreator: deselectNodeCreatorMock,
      getJurisdictionsMetadata: getMetaData(store.getState()),
      plan: {
        identifier: '2493',
        useContext: [{ code: 'interventionType', valueCodableConcept: 'IRS' }],
      } as PlanDefinition,
      rootJurisdictionId: '2942',
      selectNodeCreator: selectNodeCreatorMock,
    };
    const onJurisdictionClickMock1 = onJurisdictionClick(mockProps, () => jest.fn(), history);
    onJurisdictionClickMock1(mockMapObj, event as EventData);
    // test that selected node action is created when node is selected on map
    expect(selectNodeCreatorMock).toBeCalledTimes(1);
    expect(deselectNodeCreatorMock).toBeCalledTimes(0);
    await flushPromises();
    // test that deselect node action is created when node is selected on map
    const mockProps2 = {
      currentChildren: [
        {
          model: {
            id: '2943',
            meta: { selected: true },
          },
        },
        {
          model: {
            id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
            meta: { selected: false },
          },
        },
      ],
      currentParentNode: {
        children: [
          {
            model: {
              id: '2943',
              meta: { selected: true },
            },
          },
          {
            model: {
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              meta: { selected: false },
            },
          },
        ],
        model: {
          id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
          meta: { selected: true },
        },
      },
      deselectNodeCreator: deselectNodeCreatorMock,
      getJurisdictionsMetadata: getMetaData(store.getState()),
      plan: {
        identifier: '2493',
        useContext: [{ code: 'interventionType', valueCodableConcept: 'IRS' }],
      } as PlanDefinition,
      rootJurisdictionId: '2942',
      selectNodeCreator: selectNodeCreatorMock,
    };
    const onJurisdictionClickMock2 = onJurisdictionClick(mockProps2, () => jest.fn(), history);
    onJurisdictionClickMock2(mockMapObj, event as EventData);
    expect(selectNodeCreatorMock).toBeCalledTimes(2);
  });
  it('shows popup on hover', () => {
    const event = {
      lngLat: {
        lat: 15.065355545319008,
        lng: 101.1799767158821,
      },
      originalEvent: {},
      point: {
        x: 463.5,
        y: 477.1875,
      },
      target: {
        queryRenderedFeatures: () => {
          return [
            {
              geometry: {
                coordinates: [
                  [
                    [101.16072535514832, 15.119824869285075],
                    [101.15796539300004, 15.052626968000027],
                    [101.16026588800008, 15.052683043000059],
                    [101.16249336800007, 15.05279176700003],
                    [101.16457351200006, 15.05287588400006],
                    [101.16817184600006, 15.053010455000049],
                    [101.16992435000009, 15.05309115600005],
                    [101.16997576800009, 15.052226568000037],
                    [101.17015797000005, 15.050841579000064],
                    [101.17258921900009, 15.050799469000026],
                    [101.17424499900005, 15.050698872000055],
                    [101.17613469500009, 15.050485078000065],
                    [101.17699562400009, 15.050437232000036],
                  ],
                ],
                type: 'Polygon',
              },
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              properties: {
                geographicLevel: 3,
                name: 'Two Three Two Release Village',
                parentId: '872cc59e-0bce-427a-bd1f-6ef674dba8e2',
                status: 'Active',
                version: 0,
              },
              type: 'Feature',
            },
          ];
        },
      },
      type: 'fill',
    };

    const mockedPopup = Popup;
    const addToMock = jest.fn();
    mockedPopup.prototype.setLngLat = (e: any) => {
      expect(e).toEqual({
        lat: 15.065355545319008,
        lng: 101.1799767158821,
      });
      e.setHTML = (f: string) => {
        expect(f).toEqual(
          '<div class="jurisdiction-name"><center>Two Three Two Release Village</center></div>'
        );
        e.addTo = addToMock;
        return e;
      };
      return e;
    };

    (global as any).mapboxgl = {
      Popup: mockedPopup,
    };

    const mockStyle = {
      cursor: 'pointer',
    };

    const mockMap: any = {
      // mock unproject function
      getCanvas: jest.fn().mockImplementation(() => {
        return {
          style: {
            ...mockStyle,
          },
        };
      }),
      unproject: () => {
        return {
          lat: 15.065355545319008,
          lng: 101.1799767158821,
        };
      },
    };

    buildMouseMoveHandler(mockMap, event as any);

    expect(addToMock.mock.calls).toEqual([
      [{ getCanvas: expect.any(Function), unproject: expect.any(Function) }],
    ]);
    expect(addToMock).toBeCalledTimes(1);
  });
  it('drills down correctly based on the assignment page base URL', () => {
    const event = {
      originalEvent: {
        altKey: false,
      },
      point: {
        x: 463.5,
        y: 477.1875,
      },
      target: {
        queryRenderedFeatures: () => {
          return [
            {
              geometry: {
                coordinates: [
                  [
                    [101.16072535514832, 15.119824869285075],
                    [101.15796539300004, 15.052626968000027],
                    [101.16026588800008, 15.052683043000059],
                    [101.16249336800007, 15.05279176700003],
                    [101.16457351200006, 15.05287588400006],
                    [101.16817184600006, 15.053010455000049],
                    [101.16992435000009, 15.05309115600005],
                    [101.16997576800009, 15.052226568000037],
                    [101.17015797000005, 15.050841579000064],
                    [101.17258921900009, 15.050799469000026],
                    [101.17424499900005, 15.050698872000055],
                    [101.17613469500009, 15.050485078000065],
                    [101.17699562400009, 15.050437232000036],
                  ],
                ],
                type: 'Polygon',
              },
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              properties: {
                geographicLevel: 3,
                name: 'Two Three Two Release Village',
                parentId: '872cc59e-0bce-427a-bd1f-6ef674dba8e2',
                status: 'Active',
                version: 0,
              },
              type: 'Feature',
            },
          ];
        },
      },
    };
    const mockProps = {
      baseAssignmentURL: `${ASSIGN_PLAN_URL}/2493`,
      currentChildren: [
        {
          hasChildren: () => true,
          model: {
            id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
            meta: { selected: false },
          },
        },
      ],
      currentParentNode: {
        children: [
          {
            hasChildren: () => true,
            model: {
              id: '8fb28715-6c80-4e2c-980f-422798fe9f41',
              meta: { selected: false },
            },
          },
        ],
        model: {
          id: '2943',
          meta: { selected: false },
        },
      },
      plan: {
        identifier: '2493',
        useContext: [{ code: 'interventionType', valueCodableConcept: 'IRS' }],
      } as PlanDefinition,
      rootJurisdictionId: '2942',
    };
    const mockMapObj: any = {};
    const onJurisdictionClickMock1 = onJurisdictionClick(mockProps, () => jest.fn(), history);
    onJurisdictionClickMock1(mockMapObj, event as EventData);

    // test correct navigation for plans assignment
    expect(history.location.pathname).toEqual('/assign/2493/8fb28715-6c80-4e2c-980f-422798fe9f41');

    // test correct navigation for jurisdiction assignment
    mockProps.baseAssignmentURL = `${MANUAL_ASSIGN_JURISDICTIONS_URL}/${mockProps.plan.identifier}/${mockProps.rootJurisdictionId}`;
    const onJurisdictionClickMock2 = onJurisdictionClick(mockProps, () => jest.fn(), history);
    onJurisdictionClickMock2(mockMapObj, event as EventData);
    expect(history.location.pathname).toEqual(
      '/manualSelectJurisdictions/2493/2942/8fb28715-6c80-4e2c-980f-422798fe9f41'
    );
  });
  it('shows error div when features have invalid coordinates', async () => {
    store.dispatch(fetchTree(sampleHierarchy, '2942'));
    store.dispatch(fetchJurisdictions(fixtures.payloadWithInvalidCoordinates as any));
    store.dispatch(selectNode('2942', '2942', '2943'));
    const children = getCurrentChildren()(store.getState(), {
      currentParentId: '2942',
      leafNodesOnly: true,
      planId: '2953',
      rootJurisdictionId: '2942',
    });
    const props = {
      currentChildren: [children] as any,
      currentParentId: '3019',
      fetchJurisdictionsActionCreator: fetchJurisdictions,
      getJurisdictionsFeatures: fixtures.invalidCoordinatesGeom as any,
      plan: {
        identifier: '2493',
      } as PlanDefinition,
      rootJurisdictionId: '2942',
      serviceClass: OpenSRPService,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedAssignmentMapWrapper {...props} />
        </Router>
      </Provider>
    );
    expect(wrapper.find('div').text()).toEqual(MAP_LOAD_ERROR);
  });
});
