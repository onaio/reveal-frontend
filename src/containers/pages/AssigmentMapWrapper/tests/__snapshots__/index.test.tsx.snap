// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/AssigmentMapWrapper tests that geojson is loaded correctly in store: map renders correctly 1`] = `
<div>
  I love oov
</div>
`;
